graylog2-server/src/main/java/org/graylog2/migrations/V20220929145442_MigratePivotLimitsInViews.java

```diff
@@ -0,0 +1,149 @@
+
+                                        if (widgetIndex != null && (maxRowPivotLimit.isPresent() || maxColumnPivotLimit.isPresent())) {
+                                            widgetLimitMigrations.add(new ViewWidgetLimitMigration(viewId, queryId, widgetIndex, maxRowPivotLimit, maxColumnPivotLimit));
+                                        }
+                                    });
+                                });
+                    });
+        }
+
+        final List<WriteModel<Document>> operations = widgetLimitMigrations.stream()
+                .flatMap(widgetMigration -> {
+                    final ImmutableList.Builder<WriteModel<Document>> builder = ImmutableList.builder();
+                    widgetMigration.rowLimit().ifPresent(rowLimit -> {
+                        builder.add(updateView(widgetMigration.viewId,
+                                new Document("$set", new Document(widgetConfigPath(widgetMigration) + ".row_limit", rowLimit))
+                        ));
+                        builder.add(updateView(widgetMigration.viewId,
+                                new Document("$unset", new Document(widgetConfigPath(widgetMigration) + ".row_pivots.$[].config.limit", 1))
+                        ));
+                    });
+                    widgetMigration.columnLimit().ifPresent(columnLimit -> {
+                        builder.add(updateView(widgetMigration.viewId,
+                                new Document("$set", new Document(widgetConfigPath(widgetMigration) + ".column_limit", columnLimit))
+                        ));
+                        builder.add(updateView(widgetMigration.viewId,
+                                new Document("$unset", new Document(widgetConfigPath(widgetMigration) + ".column_pivots.$[].config.limit", 1))
+                        ));
+                    });
+                    return builder.build().stream();
+                })
+                .collect(Collectors.toList());
+
+        if (!operations.isEmpty()) {
+            this.views.bulkWrite(operations);
+        }
+
+        clusterConfigService.write(new MigrationCompleted(widgetLimitMigrations.size()));
+    }
+
+    private String widgetConfigPath(ViewWidgetLimitMigration widgetMigration) {
+        return "state." + widgetMigration.queryId() + ".widgets." + widgetMigration.widgetIndex() + ".config";
+    }
+
+    private WriteModel<Document> updateView(String viewId, Document update) {
+        return new UpdateOneModel<>(
+                new Document("_id", new ObjectId(viewId)),
+                update,
+                new UpdateOptions().upsert(false)
+        );
+    }
+
+    private Optional<Integer> extractMaxLimit(List<Document> pivots) {
+        return pivots.stream()
+                .filter(pivot -> "values".equals(pivot.get("type")))
+                .map(pivot -> pivot.get("config", Document.class))
+                .map(pivotConfig -> pivotConfig != null ? pivotConfig.getInteger("limit") : null)
```

Comment:
could be first filtered for null, too - instead of the 'pivotConfig != null`. The second filter is still needed (not that it get's deleted by accident if this suggestion is implemented)